### ISSUE TITLE
macOS: rotated JPEGs aren't handled well

### ISSUE THREAD

**Issue by dos1** [liballeg/allegro5#935]
I have a JPEG file that has a rotation tag applied to it and I display it with Allegro.

Effect on GNU/Linux (as expected):

<img src="https://dosowisko.net/odlotlinux.png" />

Effect on macOS:

<img src="https://dosowisko.net/odlotmacos.png" />

(the grain comes from an applied shader; the bitmap is slightly letterboxed)

**Comment by fatcerberus** [comment]
I'm going to guess that the red tint is because it's reading the colors in the wrong order, i.e. BGR instead of RGB, or vice versa.  Kind of strange that it honors the rotation flag but then gets the colors wrong like that.

**Comment by SiegeLord** [comment]
I'm confused. So there's two issues:

- The image is not rotated on Linux (OSX does it correctly)
- On OSX, something's off with the color channels?

**Comment by elias-pschernig** [comment]
Is it perhaps using the native image loader on OSX? I suggest we disable those by default - you really want to use libjpeg (and libpng) to load pictures, never the native loader.

**Comment by elias-pschernig** [comment]
Supporting additional formats is the original reason to have the native image loaders I guess - but the problem is how that's not portable at all. Someone unaware of the implications could develop a game on OSX and have some pictures in some strange OSX only format - and then later find out that for Windows users for their game those pictures can't load.

So I feel it would be safer to make it an explicit choice you have to enable at your own risk.

We just debugged an issue in #allegro where someone could load almost all of their .png files, except a few specific pictures resulted in garbage instead. Turns out the native OSX .png loader (or the Allegro code converting to ALLEGRO_BITMAP) can't handled some special indexed formats. Switching to the libpng loader made everything work - he just had no idea that even exists.

I don't know who builds the Macports binaries, but I would feel better if they were using libpng and libjpeg.

**Comment by elias-pschernig** [comment]
Btw. I have no idea if the JPEG issue here is related to the native loaders so this might be a bit off topic - something we should confirm first.

**Comment by guilt** [comment]
Is this an EXIF issue?

**Comment by fatcerberus** [comment]
No, the rotation is actually *honored* on OS X—the problem shown in the OP is that bitmap is rotated on load, but it then gets cropped as it tries to load the rotated image into a non-rotated surface.  It’s really bizarre.

**Comment by guilt** [comment]
Also the colors are worse. The deal is that the width and height of a jpeg are stored independently of the EXIF header. So, depending on the EXIF, one is supposed to render these into the correct angle and flip, and the final image should optionally get its width/height flipped as necessary. It is a bad bug in both cases.

**Comment by fatcerberus** [comment]
What’s weird is that switching to libjpeg apparently fixes it - which suggests a bug in the OSes’ image loaders themselves.  But that’d be a weird thing to not be picked up in testing, so maybe Allegro just isn’t using the native OS APIs correctly.

**Comment by guilt** [comment]
The ideal jpeg should show your image rotated 90 degrees (like OS X) but uncropped. Also, this image uses YCbCr so it looks like the OS X loader swapped channels or something.

**Comment by guilt** [comment]
In a noobie world, post EXIF processing, your image is 1080 x 1920 :)
